### PR TITLE
feat: (observability) Add support for OpenTelemetry traces and allow observability options to be passed. 

### PR DESCRIPTION
**Pull request by odeke-em**
This change plumbs ObservabilityConfig into Spanner, Instance and Database so that
any subsequent traces will use it when beginning spans and later on for metrics.

Updates #2079

### CODE REVIEW

File: src/index.ts
Comment (surbhigarg92) [inline src/index.ts:142]: We don't need to use "undefined" if a field is marked optional using "?"

Also please add the documentation for the new options similar to other options. 

We will also need to define "Observability Options" type

File: src/instance.ts
Comment (surbhigarg92) [inline src/instance.ts:168]: we should define observabilityConfig as optional using "?" instead of using "undefined". Same standard should be used everywhere. 

Also use param name `observabilityOptions` same as the type name.

File: src/database.ts
Comment (surbhigarg92) [inline src/database.ts:340]: should use underscore to refer this property as private. Else customer can override the property at database level . Or is there any other way we can avoid this ?

File: src/instance.ts
Comment (surbhigarg92) [reply src/instance.ts:168]: Similar to Database , use underscore to mark this property as private.

File: test/spanner.ts
Comment (surbhigarg92) [inline test/spanner.ts:5065]: Rename this test `Should use passed ObservabilityOptions in Spanner, Instance and Database"

Comment (surbhigarg92) [inline test/spanner.ts:5065]: Add one more test to verify below 
1) Register a global exporter
2) Inject a different exporter via Spanner 
3) Verify that injected exporter is used and not global.

File: src/index.ts
Comment (odeke-em) [reply src/index.ts:142]: What's the precedent on line 141 `directedReadOptions?: google.spanner.v1.IDirectedReadOptions | null;`?

ObservabilityOptions was defined in src/instrument.ts

Comment (odeke-em) [reply src/index.ts:142]: I've exposed ObservabilityOptions as a type itself.

File: src/instance.ts
Comment (odeke-em) [reply src/instance.ts:168]: Done, I've marked it as private.

File: test/spanner.ts
Comment (odeke-em) [reply test/spanner.ts:5065]: @surbhigarg92 could we please add that test after we've merged in the other code for Database? It's currently so cumbersome and walls of code just to add just an end-to-end routine for such. Currently all the PRs are so laden with tons of code and spiraling out of control.

Comment (odeke-em) [reply test/spanner.ts]: Done, thanks!

Comment (odeke-em) [reply test/spanner.ts]: I added a TODO in the tests for it. I spent time trying to plumb it through and that was wasted, meanwhile we could simply wait until we've merged the end-to-end code that uses MockSpanner and then simply have it running end-to-end and much simpler.

Comment (odeke-em) [reply test/spanner.ts]: And I've added the test, please take a look @surbhigarg92!

File: src/database.ts
Comment (odeke-em) [reply src/database.ts]: I suffixed it with an underscore so `observabilityConfig_`.

File: src/instance.ts
Comment (odeke-em) [reply src/instance.ts]: Done.

**Comment by odeke-em** [comment]
Kindly help me run the bots @surbhigarg92 @harshachinta, thank you!

File: src/database.ts
Comment (surbhigarg92) [inline src/database.ts:347]: Can you try using "private" keyword . Also it should be `_observabilityOptions` right ?

File: src/index.ts
Comment (surbhigarg92) [inline src/index.ts:142]: Add documentation for observabilityOptions above interface similar to other options

File: src/database.ts
Comment (odeke-em) [reply src/database.ts:347]: I was following the precedent in this library of `formattedName_` as I hadn't seen other variables beginning with the underscore but sure I can update that. The only problem about declaring them with the keyword private is that between classes we actually need to read and set the value and adding a setter seems like overkill. What I could do is set it to `readonly` and infer it only from the constructor as well as having set it as _observabilityOptions as you suggested.

File: src/index.ts
Comment (odeke-em) [reply src/index.ts:144]: Done, thank you @surbhigarg92!

File: src/database.ts
Comment (odeke-em) [reply src/database.ts]: So @surbhigarg92 I don't believe that reasonable customers expect to set `formattedName_` nor would a reasonable customer change the now `_observabilityOptions`, even more, making a setter or getter is unnecessary exposure.

File: src/instrument.ts
Comment (surbhigarg92) [inline src/instrument.ts:55]:
```suggestion
 * @property [enableExtendedTracing] Sets whether to enable extended OpenTelemetry tracing. Enabling this option will add the
     * following additional attributes to the traces that are generated by the client
     * db.statement: Contains the SQL statement that is being executed.
 * Alternatively, you could set environment variable `SPANNER_ENABLE_EXTENDED_TRACING=true`.
```

File: src/instrument.ts
Comment (surbhigarg92) [inline src/instrument.ts:52]:
```suggestion
 * ObservabilityOptions defines the configuration to be used for Spanner OpenTelemetry Traces.
 * @property [tracerProvider] Sets the TracerProvider to be used for traces,
 * Global TracerProvider will be used as a fallback.
```

File: src/index.ts
Comment (surbhigarg92) [inline src/index.ts:135]:
```suggestion
 * @property {ObservabilityOptions} [observabilityOptions] Sets the observability options to be used for OpenTelemetry tracing
```

**Review by surbhigarg92 — APPROVED** [review]

File: src/instrument.ts
Comment (odeke-em) [reply src/instrument.ts:55]: Thank you @surbhigarg92! Isn't hard-coding the variable defeating the purpose of prose documentation for customers or are we more like pointing them to the actual variable to look for?

File: src/index.ts
Comment (odeke-em) [reply src/index.ts]: thank you, done @surbhigarg92!

**Comment by odeke-em** [comment]
@surbhigarg92 thanks for the review! Kindly help me run the bots

File: src/instrument.ts
Comment (odeke-em) [reply src/instrument.ts]: Done and thank you @surbhigarg92! Kindly help me re-run the bots.

**Comment by odeke-em** [comment]
Updated with the latest code from main, kindly please help me re-run the bots @surbhigarg92 @rahul2393 @harshachinta.

**Pull request merged**
